contracts/royalty-group/src/contract.rs

```diff
@@ -0,0 +1,601 @@
+            info,
+            admin.map(|admin| api.addr_validate(&admin)).transpose()?,
+        )?),
+        ExecuteMsg::UpdateMembers { add, remove } => {
+            execute_update_members(deps, env, info, add, remove)
+        }
+        ExecuteMsg::AddHook { addr } => {
+            Ok(HOOKS.execute_add_hook(&ADMIN, deps, info, api.addr_validate(&addr)?)?)
+        }
+        ExecuteMsg::RemoveHook { addr } => {
+            Ok(HOOKS.execute_remove_hook(&ADMIN, deps, info, api.addr_validate(&addr)?)?)
+        }
+        ExecuteMsg::Distribute { funds } => execute_distribute_funds(deps, info, funds),
+    }
+}
+
+pub fn execute_update_members(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    add: Vec<Member>,
+    remove: Vec<String>,
+) -> Result<Response, ContractError> {
+    let attributes = vec![
+        attr("action", "update_members"),
+        attr("added", add.len().to_string()),
+        attr("removed", remove.len().to_string()),
+        attr("sender", &info.sender),
+    ];
+
+    // make the local update
+    let diff = update_members(deps.branch(), env.block.height, info.sender, add, remove)?;
+    // call all registered hooks
+    let messages = HOOKS.prepare_hooks(deps.storage, |h| {
+        diff.clone().into_cosmos_msg(h).map(SubMsg::new)
+    })?;
+    Ok(Response::new()
+        .add_submessages(messages)
+        .add_attributes(attributes))
+}
+
+// the logic from execute_update_members extracted for easier import
+pub fn update_members(
+    deps: DepsMut,
+    height: u64,
+    sender: Addr,
+    to_add: Vec<Member>,
+    to_remove: Vec<String>,
+) -> Result<MemberChangedHookMsg, ContractError> {
+    ADMIN.assert_admin(deps.as_ref(), &sender)?;
+
+    let mut total = TOTAL.load(deps.storage)?;
+    let mut diffs: Vec<MemberDiff> = vec![];
+
+    // add all new members and update total
+    for add in to_add.into_iter() {
+        let add_addr = deps.api.addr_validate(&add.addr)?;
+        MEMBERS.update(deps.storage, &add_addr, height, |old| -> StdResult<_> {
+            total -= old.unwrap_or_default();
+            total += add.weight;
+            diffs.push(MemberDiff::new(add.addr, old, Some(add.weight)));
+            Ok(add.weight)
+        })?;
+    }
+
+    for remove in to_remove.into_iter() {
+        let remove_addr = deps.api.addr_validate(&remove)?;
+        let old = MEMBERS.may_load(deps.storage, &remove_addr)?;
+        // Only process this if they were actually in the list before
+        if let Some(weight) = old {
+            diffs.push(MemberDiff::new(remove, Some(weight), None));
+            total -= weight;
+            MEMBERS.remove(deps.storage, &remove_addr, height)?;
+        }
+    }
+
+    TOTAL.save(deps.storage, &total)?;
+    Ok(MemberChangedHookMsg { diffs })
+}
+
+pub fn execute_distribute_funds(
+    deps: DepsMut,
+    info: MessageInfo,
+    funds: Coin,
+) -> Result<Response, ContractError> {
+    let messages: StdResult<Vec<_>> = MEMBERS
```

Comment:
This could get quite expensive, you can use an optimized reward distribution algorithm/contract based on the logic here: https://github.com/CosmWasm/cw-tokens/pull/8